Platform/ApollolakeBoardPkg/Library/Stage1BBoardInitLib/Stage1BBoardInitLib.c

```diff
@@ -832,15 +832,28 @@ EarlyBootDeviceInit (
 BOOLEAN
 IsFirmwareUpdate ()
 {
-  UINT16          FirmwareUpdateStatus;
-
-  //
-  // This is platform specific method. Here just use COMS.
-  //
-  IoWrite8 (0x70, FWU_BOOT_MODE_OFFSET);
-  FirmwareUpdateStatus = IoRead8 (0x71);
+  EFI_STATUS        Status;
+  UINT16            FwuReq;
+  FLASH_MAP        *FlashMapPtr;
+  FW_UPDATE_STATUS  FwUpdStatus;
+  UINT32            OffsetInBiosRgn;
+  UINT32            RsvdBase;
+
+  // Check if state machine is set to capsule processing mode.
+  FlashMapPtr = GetFlashMapPtr ();
+  Status = GetComponentInfoByPartition (FLASH_MAP_SIG_BLRESERVED, FALSE, &RsvdBase, NULL);
+  if (!EFI_ERROR (Status) && (FlashMapPtr != NULL)) {
+    OffsetInBiosRgn = FlashMapPtr->RomSize + RsvdBase;
+    ZeroMem (&FwUpdStatus, sizeof(FwUpdStatus));
+    SpiFlashRead (FlashRegionBios, OffsetInBiosRgn, sizeof(FwUpdStatus), (VOID *)&FwUpdStatus);
+    if (CheckStateMachine (&FwUpdStatus) == EFI_SUCCESS) {
+      return TRUE;
+    }
+  }

```

Comment:
Should we check only 2 methods (CMOS and scratch bit) here to know if it is firmware update mode?
After firmware update done, the flags in CMOS or scratch bit would be cleared. and it is not necessary to check firmware update machine here.

Reply:
We don't use CMOS for APL at all.   For scratch bit, it needs to be converted into state machine since the flag will be lost in case of power failure.

Reply:
oh, CMOS method was deleted in this patch.  do we need add CSME method detection here for CSME method?
With this update, it will read SPI block in every boot which should impact boot performance. Maybe we could add a PCD to disable this feature when register method is not required.

Reply:
It will only read less than 16 bytes.   So the time is very limited.   I measured the perf impact, it is less than 1ms.   So it should be fine.
Going forward, register based trigger is the standard way.    And it is also supported from ACPI level.  HECI could be optional now.

Reply:
OK. In the firmware update command, if using "-c" parameter, it would not work if here we don't check CSME method, right? so we will add it in a separate patch?

Reply:
we check both.

Reply:
OK. CSME method is checked in another function after MemoryInit.